cirq/sim/simulator.py

```diff
@@ -0,0 +1,436 @@
+                params=param_resolver,
+                measurements=measurements,
+                final_state=final_state))
+
+        return trial_results
+
+    def simulate_moment_steps(
+        self,
+        circuit: circuits.Circuit,
+        param_resolver: study.ParamResolver = None,
+        qubit_order: ops.QubitOrderOrList = ops.QubitOrder.DEFAULT,
+        initial_state: Union[int, np.ndarray] = 0,
+        extensions: extension.Extensions = None) -> Iterator['StepResult']:
+        """Returns an iterator of StepResults for each moment simulated.
+
+        Args:
+            circuit: The Circuit to simulate.
+            param_resolver: A ParamResolver for determining values of Symbols.
+            qubit_order: Determines the canonical ordering of the qubits used to
+                define the order of amplitudes in the wave function.
+            initial_state: If an int, the state is set to the computational
+                basis state corresponding to this state. Otherwise if this is
+                a np.ndarray it is the full initial state. In this case it must
+                be the correct size, be normalized (an L2 norm of 1), and
+                be safely castable to an appropriate dtype for the simulator.
+            extensions: Extensions that will be applied during the simulate.
+                See documentation of class for details.
+
+        Returns:
+            Iterator that steps through the simulation, simulating each
+            moment and returning a StepResult for each moment.
+        """
+        param_resolver = param_resolver or study.ParamResolver({})
+        return self._simulator_iterator(circuit, param_resolver, qubit_order,
+                                        initial_state, extensions)
+
+    @abc.abstractmethod
+    def _simulator_iterator(
+        self,
+        circuit: circuits.Circuit,
+        param_resolver: study.ParamResolver,
+        qubit_order: ops.QubitOrderOrList,
+        initial_state: Union[int, np.ndarray],
+        extensions: extension.Extensions
+    ) -> Iterator['StepResult']:
+        """Iterator over StepResult from Moments of a Circuit.
+
+        Args:
+            circuit: The circuit to simulate.
+            param_resolver: A ParamResolver for determining values of
+                Symbols.
+            qubit_order: Determines the canonical ordering of the qubits used to
+                define the order of amplitudes in the wave function.
+            initial_state: The full initial state. This must be the correct
+                size, be normalized (an L2 norm of 1), and be safely
+                castable to a complex type handled by the simulator.
+            extensions: Extensions that will be applied during the simulate.
+                See documentation of class for details.
+
+        Yields:
+            StepResults from simulating a Moment of the Circuit.
+        """
+        raise NotImplementedError()
+
+
+class StepResult:
```

Comment:
I think we should separate the concept of a simulator that gives you intermediate states from the concept of a simulator that gives you the final state. Some simulation strategies, e.g. path summing, cannot cheaply give the intermediate steps.

Reply:
OK, broke into two, one interface that gives full wave function at end only, and one that allows stepping.